### PR TITLE
refactoring of translatable listener

### PR DESCRIPTION
**Pull request by JarekW**
* Tests are unchanged, they should pass :)
* No remove/change in interfaces of classes, 1 new public method.
* To reduce number of parameters translatable ClassMetadata has collection of `TranslationAssociationMetadata` that holds information about single association to translatable entities.
* `getTranslations` method in `TranslatableRepository` was protected. I've change it to public, because same functionality is duplicated in `TranslatableListener` and repository has similar method `findTranslation`
* Also to reduce parameters passing `ObjectAssociationContext` now holds information about object and its association (only one) to translation.
* `TranslationHelper` deal with all the copying/clearing of properties from entity and it's translation

What do you think? Please review

### CODE REVIEW

File: lib/FSi/DoctrineExtensions/Translatable/ObjectAssociationContext.php [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+        return $repository;
+    }
+
+    /**
+     * @return \Symfony\Component\PropertyAccess\PropertyAccessor
+     */
+    private function getPropertyAccessor()
+    {
+        if (!isset($this->propertyAccessor)) {
+            $this->propertyAccessor = PropertyAccess::createPropertyAccessor();
```
Comment (chives) [inline lib/FSi/DoctrineExtensions/Translatable/ObjectAssociationContext.php]: I would consider passing one instance of `PropertyAccessor` to every instance of `ObjectAssociationContext`

File: lib/FSi/DoctrineExtensions/Translatable/ObjectAssociationContext.php [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+namespace FSi\DoctrineExtensions\Translatable;
+
+use Doctrine\Common\Persistence\Mapping\ClassMetadata;
+use Doctrine\Common\Persistence\ObjectManager;
+use FSi\DoctrineExtensions\Translatable\Mapping\TranslationAssociationMetadata;
+use FSi\DoctrineExtensions\Translatable\Model\TranslatableRepositoryInterface;
+use Symfony\Component\PropertyAccess\PropertyAccess;
+use Symfony\Component\PropertyAccess\PropertyAccessor;
+
+class ObjectAssociationContext
```
Comment (chives) [inline lib/FSi/DoctrineExtensions/Translatable/ObjectAssociationContext.php]: Maybe `ObjectTranslationContext` would be more suitable name

**Comment by chives** [comment]
I think that `TranslationHelper` and `ObjectAssociationContext`/`ObjectTranslationContext` should be marked as `@Internal`

**Comment by JarekW** [comment]
@chives, context class changed. You can check last commit. It's not yet cached in listener, but you can check if increase in parameters is ok.

**Comment by JarekW** [comment]
@fsi-open how about now? :)

File: lib/FSi/DoctrineExtensions/Translatable/TranslatableListener.php [hunk 3]
```diff
@@ -342,250 +311,68 @@ private function updateObjectTranslations(ObjectManager $objectManager, $object)
...
-    private function clearObjectProperties(ObjectManager $objectManager, $object, $translationAssociation)
-    {
-        $translatableProperties = $this->getTranslatableMetadata($objectManager, $object)->getTranslatableProperties();
-        $translationMeta = $this->getTranslationClassMetadata($objectManager, $object, $translationAssociation);
-        foreach ($translatableProperties[$translationAssociation] as $property => $translationField) {
-            if ($translationMeta->isCollectionValuedAssociation($translationField)) {
-                $this->getPropertyAccessor()->setValue($object, $property, array());
+            if ($hasTranslatedProperties) {
+                $this->translationHelper->copyPropertiesToTranslation(
+                    $objectManager,
```
Comment (chives) [inline lib/FSi/DoctrineExtensions/Translatable/TranslatableListener.php]: `$objectManager` get be retrieved from `$context` and does not need to be passed directly

File: lib/FSi/DoctrineExtensions/Translatable/TranslatableListener.php [hunk 4]
```diff
@@ -342,250 +311,68 @@ private function updateObjectTranslations(ObjectManager $objectManager, $object)
...
+                    $context,
+                    $object,
+                    $locale
+                );
             } else {
-                $this->getPropertyAccessor()->setValue($object, $property, null);
+                $this->translationHelper->removeEmptyTranslation(
+                    $objectManager,
```
Comment (chives) [inline lib/FSi/DoctrineExtensions/Translatable/TranslatableListener.php]: `$objectManager` get be retrieved from `$context` and does not need to be passed directly

**Comment by JarekW** [comment]
@fsi-open

File: lib/FSi/DoctrineExtensions/Translatable/TranslatableListener.php [hunk 5]
```diff
@@ -342,250 +311,66 @@ private function updateObjectTranslations(ObjectManager $objectManager, $object)
...
-        $translations = $this->getPropertyAccessor()->getValue($object, $translationAssociation);
-        if (!isset($translations)) {
-            $translations = new ArrayCollection();
+        if (empty($this->classTranslationContexts[$className][$associationName])) {
+            $context = new ClassTranslationContext($objectManager, $classMeta, $associationMeta);
+            $this->classTranslationContexts[$className][$associationName] = $context;
         }
 
-        return $translations;
-    }
```
Comment (chives) [inline lib/FSi/DoctrineExtensions/Translatable/TranslatableListener.php]: one useless empty line

**Comment by chives** [comment]
beside that :+1:

**Comment by JarekW** [comment]
@fsi-open

File: lib/FSi/DoctrineExtensions/Translatable/TranslationHelper.php [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+    {
+        $translationMeta = $context->getTranslationMetadata();
+        $properties = $context->getAssociationMetadata()->getProperties();
+        $propertyAccessor = $this->getPropertyAccessor();
+
+        foreach ($properties as $property => $translationField) {
+            $value = $propertyAccessor->getValue($object, $property);
+            if ($translationMeta->isCollectionValuedAssociation($translationField) && count($value)) {
+                return true;
+            } elseif (!$translationMeta->isCollectionValuedAssociation($translationField) && (null !== $value)) {
```
Comment (szymach) [inline lib/FSi/DoctrineExtensions/Translatable/TranslationHelper.php]: could be in one `if`

**Comment by szymach** [comment]
aside from that :+1:

**Comment by JarekW** [comment]
@rn0 :)

**Comment by JarekW** [comment]
@rn0

**Comment by rn0** [comment]
:+1:

**Pull request merged**